nixos/modules/services/networking/lokinet.nix

```diff
@@ -0,0 +1,219 @@
+        <link xlink:href="https://docs.loki.network/Lokinet/Guides/LokinetConfig/"/>
+        for details on supported values;
+      '';
+    };
+
+    threads = mkOption {
+      type = types.int;
+      default = 4;
+      example = 8;
+      description = "The number of crypto worker threads.";
+    };
+
+    nickname = mkOption {
+      type = with types; nullOr str;
+      default = null;
+      example = "lokinet";
+      description = "Sets router nickname to what you want.";
+    };
+
+    minRouters = mkOption {
+      type = types.int;
+      default = 4;
+      example = 6;
+      description = "Maintain minimum connections to other routers.";
+    };
+
+    maxRouters = mkOption {
+      type = types.int;
+      default = 6;
+      example = 8;
+      description = "Hard limit of routers globally we are connected to at any given time.";
+    };
+
+    loggingType = mkOption {
+      type = with types; nullOr str;
+      default = null;
+      example = "syslog";
+      description = "Logging method to use.";
+    };
+
+    loggingFile = mkOption {
+      type = types.path;
+      default = "/var/lib/lokinet/log";
+      example = "/home/okina/lokinet/log";
+      description = "Path to the lokinet log, only enabled if loggingType is set to \"file\".";
+    };
+
+    resolverUpstream = mkOption {
+      type = types.str;
+      default = "1.1.1.1";
+      example = "8.8.8.8";
+      description = "DNS resolver upstream address.";
+    };
+
+    resolverBind = mkOption {
+      type = types.str;
+      default = "127.3.2.1:53";
+      example = "127.3.2.1:52";
+      description = "DNS resolver bind address.";
+    };
+
+    strictConnect = mkOption {
+      type = types.bool;
+      default = false;
+      description = "Add router with pubkey to list of routers we connect directly to.";
+    };
+
+    exitNode = mkOption {
+      type = types.bool;
+      default = false;
+      description = "Use router with pubkey as an exit node";
+    };
+
+    snapp = mkOption {
+      default = null;
+      description = "Snapp configuration.";
+
+      type = types.nullOr (types.submodule {
+        options = {
+          ifAddr = mkOption {
+            type = types.str;
+            default = "10.33.0.1/16";
+            example = "10.34.0.1/16";
+            description = "The ip range to allocate to this snapp.";
+          };
+
+          ifName = mkOption {
+            type = types.str;
+            default = "snapp-tun0";
+            example = "snapp-tun1";
+            description = "The name to try and give to the network interface this snapp owns.";
+          };
+        };
+      });
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd = {
+      services.lokinet = {
+        description = "lokinet";
+        after = [ "network-online.target" ];
+        wantedBy = [ "multi-user.target" ];
+        preStart = "${pkgs.curl}/bin/curl https://seed.lokinet.org/bootstrap.signed -o ${dataDir}/bootstrap.signed";
```

Comment:
```
preStart = ''
  cp -f ${configFile} $STATE_DIRECTORY/lokinet.ini
  cp -f ${snappFile} $STATE_DIRECTORY/snapp-example.ini
  ${pkgs.curl}/bin/curl https://seed.lokinet.org/bootstrap.signed > ${dataDir}/bootstrap.signed
'';
```
You can't use `tmpfiles.d` with `DynamicUser` as the user won't exist when `tmpfiles.d` runs. When your `lokinet` service starts it sees an existing directory that it knows wasn't previously managed by `DynamicUser` mechanisms, it also sees a wonky user ownership as the `lokinet` user doesn't properly exist when the directory was created... which leads to the error you saw.

Fortunately specifying `StateDirectory` takes care of directory creation for you, so you don't actually need `tmpfiles.d`.

**Note:** I don't think `curl` overwrites by default, but I haven't checked that so I switched the code to overwrite. If passing `-o` to `curl` does overwrite feel free to ignore that suggestion.

Reply:
I still get this issue when using
```
preStart = ''
  mkdir -p ${escapeShellArg dataDir}/.lokinet
  cd ${escapeShellArg dataDir}
  ln -sf ${configFile} lokinet.ini
  ln -sf ${snappFile} snapp-example.ini
  ${pkgs.curl}/bin/curl https://seed.lokinet.org/bootstrap.signed -o ${dataDir}/bootstrap.signed
'';
```

Reply:
Same with your variation that I just noticed and
```
preStart = ''
  mkdir -p $STATE_DIRECTORY/.lokinet
  cp -f ${configFile} $STATE_DIRECTORY/.lokinet/lokinet.ini
  cp -f ${snappFile} $STATE_DIRECTORY/.lokinet/snapp-example.ini
  ${pkgs.curl}/bin/curl https://seed.lokinet.org/bootstrap.signed -o $STATE_DIRECTORY/.lokinet/bootstrap.signed
'';
```

Reply:
@Chiiruno try:
- `systemd.services.lokinet.environment.HOME = "/var/lib/lokinet";`
- `ExecStart = "${pkgs.lokinet}/bin/lokinet --verbose";` as this appears to provide valuable debugging information
- `$STATE_DIRECTORY/.lokinet` -> `${dataDir}`

I was able to get past the error you showed. You likely need to `sudo rm -rf /var/lib/private/lokinet` and `sudo rm -rf /var/lib/lokinet` first, though.